File: OrchextraApp/Carthage/Checkouts/gigigo-ios-lib/Source/GIGUtils/Date/NSDate+GIGExtension.m
//
//  NSDate+GIGExtension.m
//  giglibrary
//
//  Created by Sergio Baró on 16/09/2013.
//  Copyright (c) 2013 Gigigo. All rights reserved.
//

#import "NSDate+GIGExtension.h"
#import <GIGLibrary/GIGLibrary-Swift.h>


@implementation NSDate (GIGExtension)


+ (NSDate *)dateFromString:(NSString *)dateString format:(NSString *)format locale:(NSLocale *)locale
{
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    dateFormatter.dateFormat = format;
    
    if (locale)
    {
        dateFormatter.locale = locale;
    }
    
    return [dateFormatter dateFromString:dateString];
}

+ (NSDate *)randomDate
{
    NSInteger day = arc4random_uniform(31) + 1;
    NSInteger month = arc4random_uniform(12) + 1;
    NSInteger year = arc4random_uniform(2013 - 2000) + 2000;
    
    NSInteger hour = arc4random_uniform(24) + 1;
    NSInteger minutes = arc4random_uniform(60) + 1;
    NSInteger seconds = arc4random_uniform(60) + 1;
    
    NSString *dateString = [NSString stringWithFormat:@"%02d/%02d/%04d %02d:%02d:%02d", (int)day, (int)month, (int)year, (int)hour, (int)minutes, (int)seconds];
	
	return [NSDate dateFromString:dateString format:@"dd/MM/yyyy HH:mm:ss" locale:[NSLocale currentLocale]];
}

+ (NSTimeInterval)timeIntervalSince1970
{
    return [NSDate timeIntervalSinceReferenceDate] + NSTimeIntervalSince1970;
}

+ (NSDate *)yesterday
{
    return [[NSDate date] dateByAddingDays:-1];
}

+ (NSDate *)afterYesterday
{
    return [[NSDate date] dateByAddingDays:-2];
}

+ (NSDate *)tomorrow
{
    return [[NSDate date] dateByAddingDays:1];
}

- (NSString *)stringWithFormat:(NSString *)dateFormat locale:(NSLocale *)locale
{
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
	dateFormatter.dateFormat = dateFormat;
    dateFormatter.AMSymbol = @"";
    dateFormatter.PMSymbol = @"";
	
    if (locale)
    {
        dateFormatter.locale = locale;
    }
    
	return [dateFormatter stringFromDate:self];
}

- (NSString *)stringWithFormat:(NSString *)dateFormat
{
    return [self stringWithFormat:dateFormat locale:nil];
}

- (BOOL)isToday
{
    return ([self compareWithoutTimeWithDate:[NSDate date]]);
}

- (BOOL)isYesterday
{
    return ([self compareWithoutTimeWithDate:[NSDate yesterday]]);
}

- (BOOL)compareToSecondsWithDate:(NSDate *)date
{
    return ([[self dateWithSeconds] isEqualToDate:[date dateWithSeconds]]);
}

- (NSInteger)numberOfDaysFromNow
{
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *components = [calendar components:NSCalendarUnitDay
                                                        fromDate:[NSDate date]
                                                          toDate:self
                                                         options:kNilOptions];
    
    return [components day];
}

- (NSDate *)dateByAddingDays:(NSInteger)numberOfDays
{
    return [self dateByAddingTimeInterval:60*60*24*numberOfDays];
}

#pragma mark - Private

- (BOOL)compareWithoutTimeWithDate:(NSDate *)date
{
    return ([[self dateWithoutTime] isEqualToDate:[date dateWithoutTime]]);
}

- (NSDate *)dateWithSeconds
{
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *selfComponents = [calendar components:(NSCalendarUnitEra|NSCalendarUnitYear|NSCalendarUnitMonth|NSCalendarUnitDay|NSCalendarUnitHour|NSCalendarUnitMinute|NSCalendarUnitSecond)
                                                   fromDate:self];
    
    return [calendar dateFromComponents:selfComponents];
}

- (NSDate *)dateWithoutTime
{
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *selfComponents = [calendar components:(NSCalendarUnitEra|NSCalendarUnitYear|NSCalendarUnitMonth|NSCalendarUnitDay)
                                                   fromDate:self];
    
    return [calendar dateFromComponents:selfComponents];
}

@end

File: OrchextraApp/Carthage/Checkouts/gigigo-ios-lib/Source/GIGUtils/Date/NSDate+GIGExtension.swift
//
//  NSDate+AlwaysOn.swift
//  AlwaysOn
//
//  Created by Alejandro Jiménez Agudo on 22/2/16.
//  Copyright © 2016 Gigigo S.L. All rights reserved.
//

import Foundation


public let DateISOFormat = "yyyy-MM-dd'T'HH:mm:ssZ"


/**
Error type for invalid values

Cases:
* InvalidHour: Throws when the hour set is not in 0..24 range
* InvalidMinutes: Throws when the hour set is not in 0..60 range
* InvalidSeconds: Throws when the hour set is not in 0..60 range

- Author: Alejandro Jiménez
- Since: 1.1.3
*/
public enum ErrorDate: Error {
	case invalidHour
	case invalidMinutes
	case invalidSeconds
}

public extension Date {
	
	/// Date from string with ISO format.
	public static func dateFromString(_ dateString: String, format: String = DateISOFormat) -> Date? {
		let dateFormat = DateFormatter()
		dateFormat.dateFormat = format

		let date = dateFormat.date(from: dateString)
		return date
	}
	
	public static func today() -> Date {
		return Date()
	}
	
	public func string(with format: String = DateISOFormat) -> String {
		let dateFormatter = DateFormatter()
		dateFormatter.dateFormat = format
		dateFormatter.amSymbol = ""
		dateFormatter.pmSymbol = ""
		return dateFormatter.string(from: self)
	}
	
	
	/**
	Set the time to a NSDate
	
	- parameters:
		- hour: The hour to be set
		- minutes: The minutes to be set. Optional, 0 by default
		- seconds: The seconds to be set. Optional, 0 by default
	
	- important: The time is set in local time respecting the user time zone.
	Examples (in Spain):
	* setHour(14) to a summer date (UTC+2) returns -> 12:00:00 +0000
	* setHour(14) to a winter date (UTC+1) returns -> 13:00:00 +0000
	
	- throws: An error of type ErrorDate
	- returns: A new NSDate with the same date and the time set.
	- Author: Alejandro Jiménez
	- Since: 1.1.3
	*/
	public func setHour(_ hour: Int, minutes: Int = 0, seconds: Int = 0) throws -> Date {
		guard 0 <= hour && hour < 24		else { throw ErrorDate.invalidHour }
		guard 0 <= minutes && minutes < 60	else { throw ErrorDate.invalidMinutes }
		guard 0 <= seconds && seconds < 60	else { throw ErrorDate.invalidSeconds }
		
		let calendar = Calendar(identifier: Calendar.Identifier.gregorian)
		var components = (calendar as NSCalendar).components(([.day, .month, .year]), from: self)
		components.hour = hour
		components.minute = minutes
		components.second = seconds
		
		return calendar.date(from: components)!
	}
}


/// Add days to a date
public func +(date: Date, days: Int) -> Date {
	let newDate = (date as NSDate).addingDays(days)
	return newDate!
}

/// Substract days to a date
public func -(date: Date, days: Int) -> Date {
	let newDate = date + (-days)
	return newDate
}


//public func >(lhs: Date, rhs: Date) -> Bool {
//	let result = lhs.compare(rhs)
//	
//	switch result {
//	case .orderedDescending:
//		return true
//	
//	default:
//		return false
//	}
//}
//
//public func <(lhs: Date, rhs: Date) -> Bool {
//	let result = lhs.compare(rhs)
//	
//	switch result {
//	case .orderedAscending:
//		return true
//		
//	default:
//		return false
//	}
//}
//
//public func ==(lhs: Date, rhs: Date) -> Bool {
//	let result = lhs.compare(rhs)
//	
//	switch result {
//	case .orderedSame:
//		return true
//		
//	default:
//		return false
//	}
//}
